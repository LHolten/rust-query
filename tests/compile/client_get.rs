use rust_query::{Snapshot, Value};

fn main() {}

fn test<'a, S>(db: &Snapshot<'a, S>, val: impl Value<'a, S, Typ = i64>) {
    db.get(val);
}
