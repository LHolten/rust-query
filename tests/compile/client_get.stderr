error: lifetime may not live long enough
 --> tests/compile/client_get.rs:6:5
  |
5 | fn test<'a, S>(db: &Snapshot<'a, S>, val: impl Value<'a, S, Typ = i64>) {
  |         -- lifetime `'a` defined here
6 |     db.get(val);
  |     ^^^^^^^^^^^ requires that `'a` must outlive `'static`
  |
note: due to current limitations in the borrow checker, this implies a `'static` lifetime
 --> src/transaction.rs
  |
  |     pub fn get<'s, T>(&'s self, val: impl for<'x> FromRow<'x, 's, S, Out = T>) -> T {
  |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: implementation of `FromRow` is not general enough
 --> tests/compile/client_get.rs:6:5
  |
6 |     db.get(val);
  |     ^^^^^^^^^^^ implementation of `FromRow` is not general enough
  |
  = note: `impl Value<'a, S, Typ = i64>` must implement `FromRow<'0, '_, S>`, for any lifetime `'0`...
  = note: ...but it actually implements `FromRow<'1, '_, S>`, for some specific lifetime `'1`

error[E0308]: mismatched types
 --> tests/compile/client_get.rs:6:5
  |
6 |     db.get(val);
  |     ^^^^^^^^^^^ one type is more general than the other
  |
  = note: expected associated type `<impl Value<'a, S, Typ = i64> as rust_query::Value<'_, S>>::Typ`
             found associated type `<impl Value<'a, S, Typ = i64> as rust_query::Value<'x, S>>::Typ`
note: the lifetime requirement is introduced here
 --> src/transaction.rs
  |
  |     pub fn get<'s, T>(&'s self, val: impl for<'x> FromRow<'x, 's, S, Out = T>) -> T {
  |                                                                      ^^^^^^^
